db/migrations/000001_db_setup.up.sql

```diff
@@ -0,0 +1,178 @@
+CREATE TYPE
+    "order_status" AS ENUM (
+        'pending',
+        'paid',
+        'shipped',
+        'delivered',
+        'cancelled'
+    );
+
+CREATE TYPE
+    "coupon_type" AS ENUM (
+        'percentage',
+        'fixed',
+        'shipping'
+    );
+
+CREATE TYPE "role_type" AS ENUM ( 'admin', 'customer' );
+
+CREATE TABLE
+    "cart" (
+        "id" SERIAL PRIMARY KEY,
+        "user_id" INT NOT NULL,
+        "shop_id" INT NOT NULL
+    );
+
+CREATE TABLE
+    "cart_product" (
+        "cart_id" INT NOT NULL,
+        "product_id" INT NOT NULL,
+        "quantity" INT NOT NULL
+    );
+
+CREATE TABLE
+    "order_history" (
+        "id" SERIAL PRIMARY KEY,
+        "user_id" INT NOT NULL,
+        "shop_id" INT NOT NULL,
+        "shipment" INT NOT NULL,
+        "total_price" INT NOT NULL,
+        "status" order_status NOT NULL,
+        "created_at" TIMESTAMPTZ NOT NULL DEFAULT NOW()
+    );
+
+CREATE TABLE
+    "order_detail" (
+        "order_id" INT NOT NULL,
+        "product_id" INT NOT NULL,
+        "product_version" INT NOT NULL,
+        "quantity" INT NOT NULL
+    );
+
+CREATE TABLE
+    "cart_coupon" (
+        "cart_id" INT NOT NULL,
+        "coupon_id" INT NOT NULL
+    );
+
+CREATE TABLE
+    "product" (
+        "id" SERIAL PRIMARY KEY,
+        "version" SERIAL,
```

Comment:
When editing a product, we will add a new row to the table. 
SQL will look like this.
```sql
INSERT INTO "product" ("version", "shop_id", "name", "description", "price", "image_id", "due_date", "stock", "sales", "enabled")
SELECT
    p."version" + 1,
    p."shop_id",
    'New Name', -- Replace with the new name
    'New Description', -- Replace with the new description
    99.99, -- Replace with the new price
    p."image_id",
    now(), -- Replace with the new due date
    50, -- Replace with the new stock
    0, -- Reset sales to 0 for the new version
    TRUE -- Enable the new version
FROM "product" p
WHERE p."id" = 123; -- Replace 123 with the product ID you are updating

```

It seems like we can't use the serial number to auto-increment the version. Should we consider changing it to an unsigned integer?

Reply:
shouldn't you just leave it alone and let it handle itself?

Reply:
If my memory is correct, the serial column will auto-increment by the largest number in that column, but the our version should be increased of the last product version.